src/transformers/trainer.py

```diff
@@ -579,20 +590,37 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
             # Reinitializes optimizer and scheduler
             self.optimizer, self.lr_scheduler = None, None
 
+        # Keeping track whether we can can len() on the dataset or not
+        train_dataset_is_sized = isinstance(self.train_dataset, collections.abc.Sized)
+
         # Data loader and number of training steps
         train_dataloader = self.get_train_dataloader()
-        num_update_steps_per_epoch = len(train_dataloader) // self.args.gradient_accumulation_steps
```

Comment:
I think regrouping all in the tests below make the code less readable. So this should stay here with an `if train_dataset_is_sized` with the corresponding block else, then the next step stays roughly the same (just sets `num_train_epochs` to 1 if necessary and removes the assert.

Reply:
Re-arranged in a different way, and added comments about who's who.
Checked that whatever execution path, all 3 variables get defined.